### PR TITLE
Add DSP Panic dump support for AMD platform

### PR DESCRIPTION
**Pull request by reddysujith**
This Patch  series consists of
>>Use poll function instead of single read ACP_SHA_DSP_FW_QUALIFIER
>>Add support for IPC and DSP dumps

### CODE REVIEW

File: sound/soc/sof/amd/acp.c [hunk 1]
```diff
@@ -255,8 +255,10 @@ int configure_and_run_sha_dma(struct acp_dev_data *adata, void *image_addr,
 	if (ret)
 		return ret;
 
-	fw_qualifier = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_SHA_DSP_FW_QUALIFIER);
-	if (!(fw_qualifier & DSP_FW_RUN_ENABLE)) {
+	ret = snd_sof_dsp_read_poll_timeout(sdev, ACP_DSP_BAR, ACP_SHA_DSP_FW_QUALIFIER,
+					    fw_qualifier, fw_qualifier & DSP_FW_RUN_ENABLE,
+					    ACP_REG_POLL_INTERVAL, ACP_DMA_COMPLETE_TIMEOUT_US);
+	if (ret < 0) {
```
Comment (plbossart) [inline sound/soc/sof/amd/acp.c:261]: commit message grammar?

Use the snd_sof_dsp_read_poll_timeout function to successfully read the FW_QUALIFIER register

File: sound/soc/sof/amd/acp-common.c [hunk 2]
```diff
@@ -33,6 +34,107 @@ int acp_dai_probe(struct snd_soc_dai *dai)
 }
 EXPORT_SYMBOL_NS(acp_dai_probe, SND_SOC_SOF_AMD_COMMON);
 
+/** amd_sof_ipc_dump():This function is called when IPC tx timed out.
```
Comment (plbossart) [inline sound/soc/sof/amd/acp-common.c]: bad kernel-doc formatting

````c
/** 
 * amd_sof_ipc_dump() - This function is called when IPC tx timed out.
 * 
````

File: sound/soc/sof/amd/acp-common.c [hunk 3]
```diff
@@ -33,6 +34,107 @@ int acp_dai_probe(struct snd_soc_dai *dai)
...
+	u32 dsp_msg, dsp_ack, host_msg, host_ack, irq_stat;
+
+	dsp_msg = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + dsp_msg_write);
+	dsp_ack = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + dsp_ack_write);
+	host_msg = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + host_msg_write);
+	host_ack = snd_sof_dsp_read(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + host_ack_write);
+	irq_stat = snd_sof_dsp_read(sdev, ACP_DSP_BAR, base + DSP_SW_INTR_STAT_OFFSET);
+
+	dev_err(sdev->dev,
+		"dsp_msg = 0x%x dsp_ack = 0x%x host_msg = 0x%x host_ack = 0x%x irq_stat = 0x%x\n",
```
Comment (plbossart) [inline sound/soc/sof/amd/acp-common.c]: %#x

**Review by plbossart — CHANGES_REQUESTED** [review]
couple of cosmetic issues

File: sound/soc/sof/amd/acp-common.c [hunk 4]
```diff
@@ -33,6 +34,108 @@ int acp_dai_probe(struct snd_soc_dai *dai)
 }
 EXPORT_SYMBOL_NS(acp_dai_probe, SND_SOC_SOF_AMD_COMMON);
 
+/**
+ * amd_sof_ipc_dump():This function is called when IPC tx timed out.
```
Comment (plbossart) [inline sound/soc/sof/amd/acp-common.c]: kernel doc needs a hyphen, not a semicolumn

amd_sof_ipc_dump() - This function is called when IPC tx timed out.

File: sound/soc/sof/amd/acp-common.c (hunk 2 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-common.c]: Done. Thanks

File: sound/soc/sof/amd/acp-common.c (hunk 3 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-common.c]: Done Thanks

File: sound/soc/sof/amd/acp.c (hunk 1 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp.c:261]: Done ,Thanks

File: sound/soc/sof/amd/acp-ipc.c [hunk 5]
```diff
@@ -165,7 +173,7 @@ irqreturn_t acp_sof_ipc_irq_thread(int irq, void *context)
 	if (dsp_msg) {
 		snd_sof_ipc_msgs_rx(sdev);
 		acp_dsp_ipc_host_done(sdev);
-		ipc_irq = true;
+		return IRQ_HANDLED;
```
Comment (plbossart) [inline sound/soc/sof/amd/acp-ipc.c]: this changes the flow. With the previous code, you would have tried to deal with the dsp_ack. That sounds like a different change that needs to go in a different patch, if that was intentional indeed.

File: sound/soc/sof/amd/acp-ipc.c [hunk 6]
```diff
@@ -177,11 +185,18 @@ irqreturn_t acp_sof_ipc_irq_thread(int irq, void *context)
...
+	}
+
+	acp_mailbox_read(sdev, sdev->debug_box.offset, &status, sizeof(u32));
+	if ((status & SOF_IPC_PANIC_MAGIC_MASK) == SOF_IPC_PANIC_MAGIC) {
+		bool non_recoverable = true;
+
+		snd_sof_dsp_panic(sdev, sdev->dsp_oops_offset, non_recoverable);
+		return IRQ_HANDLED;
```
Comment (plbossart) [inline sound/soc/sof/amd/acp-ipc.c:193]: it's also surprising that there are two calls to 'panic' reading from two different places. That's worthy of a comment if that again was intentional.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/amd/acp-ipc.c (hunk 5 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-ipc.c]: Here we are handling one to one communication from DSP to HOST in synchronous manner. we never get two conditions met, so we optimized and done.

Comment (plbossart) [reply sound/soc/sof/amd/acp-ipc.c]: separate change then. this has nothing to do with the panic stuff

File: sound/soc/sof/amd/acp-ipc.c (hunk 6 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-ipc.c:193]: Yes, 


Case 1:Getting panic before  FW_READY .

Here DSP and HOST sides have different mailbox offsets and also exception box offset was not set. there is only one common offset is DSP mailbox offset. So reading panic status from DSP box first four bytes and after that (XOOPS,PANIC,STACK ) info in DSP mailbox.

Case 2:Getting panic After FW_READY ipc from DSP.

The First message from DSP is FW_READY. In the  FW_READY we will get window information (dsp, host, debug, exception box all offsets and sizes are set that will keep offsets same for both DSP and HOST).

File: sound/soc/sof/amd/acp-common.c (hunk 4 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-common.c]: Done ,Thanks

File: sound/soc/sof/amd/acp-ipc.c (hunk 5 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-ipc.c]: Changes Done ..Optimization patch will be done later.

Only added code related to DSP panic Stuff.

File: sound/soc/sof/amd/acp.c [hunk 7]
```diff
@@ -255,8 +255,10 @@ int configure_and_run_sha_dma(struct acp_dev_data *adata, void *image_addr,
...
 		dev_err(sdev->dev, "PSP validation failed\n");
 		return -EINVAL;
```
Comment (ranj063) [inline sound/soc/sof/amd/acp.c]: return ret to capture the error properly

File: sound/soc/sof/amd/acp-common.c [hunk 8]
```diff
@@ -33,6 +34,108 @@ int acp_dai_probe(struct snd_soc_dai *dai)
 }
 EXPORT_SYMBOL_NS(acp_dai_probe, SND_SOC_SOF_AMD_COMMON);
 
+/**
+ * amd_sof_ipc_dump() - This function is called when IPC tx timed out.
```
Comment (ranj063) [inline sound/soc/sof/amd/acp-common.c]: typo: "times out" instead of timed out

File: sound/soc/sof/amd/acp-common.c [hunk 9]
```diff
@@ -33,6 +34,108 @@ int acp_dai_probe(struct snd_soc_dai *dai)
...
+		 */
+		acp_mailbox_read(sdev, sdev->dsp_box.offset, &status, sizeof(u32));
+		sdev->dsp_oops_offset = sdev->dsp_box.offset + sizeof(status);
+	}
+
+	/* Get information about the registers, the filename and line
+	 * number and the stack.
+	 */
+	amd_get_registers(sdev, &xoops, &panic_info, stack,
+			  AMD_STACK_DUMP_SIZE);
```
Comment (ranj063) [inline sound/soc/sof/amd/acp-common.c]: can this be on one line?

File: sound/soc/sof/amd/acp-ipc.c [hunk 10]
```diff
@@ -154,8 +154,17 @@ irqreturn_t acp_sof_ipc_irq_thread(int irq, void *context)
 				     offsetof(struct scratch_ipc_conf, sof_dsp_ack_write);
 	bool ipc_irq = false;
 	int dsp_msg, dsp_ack;
+	unsigned int status;
 
 	if (sdev->first_boot && sdev->fw_state != SOF_FW_BOOT_COMPLETE) {
+		acp_mailbox_read(sdev, sdev->dsp_box.offset, &status, sizeof(status));
+		if ((status & SOF_IPC_PANIC_MAGIC_MASK) == SOF_IPC_PANIC_MAGIC) {
+			bool non_recoverable = true;
```
Comment (ranj063) [inline sound/soc/sof/amd/acp-ipc.c]: why do you need this variable? Why not pass true directly? same comment below as well

File: sound/soc/sof/amd/acp.c (hunk 7 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp.c]: Done ,Thanks

File: sound/soc/sof/amd/acp-common.c (hunk 9 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-common.c]: Done ,Thanks

File: sound/soc/sof/amd/acp-ipc.c (hunk 10 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-ipc.c]: Yes we can remove .Done ,Thanks

File: sound/soc/sof/amd/acp-common.c (hunk 8 above)
Comment (reddysujith) [reply sound/soc/sof/amd/acp-common.c]: Done ,Thanks

**Review by plbossart — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Review by bhiregoudar — APPROVED** [review]

**Review by bhiregoudar — APPROVED** [review]

**Review by sunilkumardommati — APPROVED** [review]
